manager/allocator/networkallocator/portallocator.go

```diff
@@ -91,40 +139,63 @@ func reconcilePortConfigs(s *api.Service) []*api.PortConfig {
 		return s.Spec.Endpoint.Ports
 	}
 
-	allocatedPorts := make(map[api.PortConfig]*api.PortConfig)
+	portStates := allocatedPorts{}
 	for _, portState := range s.Endpoint.Ports {
 		if portState.PublishMode != api.PublishModeIngress {
 			continue
 		}
-
-		allocatedPorts[getPortConfigKey(portState)] = portState
+		portStates.addState(portState)
 	}
 
 	var portConfigs []*api.PortConfig
+
+	// Process the portConfig with portConfig.PublishMode != api.PublishModeIngress
+	// Iterate portConfigs with PublishModeIngress
 	for _, portConfig := range s.Spec.Endpoint.Ports {
 		// If the PublishMode is not Ingress simply pick up
 		// the port config.
 		if portConfig.PublishMode != api.PublishModeIngress {
 			portConfigs = append(portConfigs, portConfig)
 			continue
```

Comment:
Thanks @aaronlehmann. I take a look and realized that the first loop and the second loop could be merged into one loop, as `PublishMode != PublishModeIngress` and `PublishedPort != 0` does not interference with each other. So now the loops has been reduced from 3 to 2 (`continue` has been kept because of this).
(The processing of `PublishedPort == 0` still has to be separate).